notification.c

```diff
@@ -352,15 +357,46 @@ void notification_execute_binding(struct mako_notification *notif,
 		struct mako_action *action;
 		wl_list_for_each(action, &notif->actions, link) {
 			if (strcmp(action->key, DEFAULT_ACTION_KEY) == 0) {
-				char *activation_token = create_xdg_activation_token(
-					ctx->surface, ctx->seat, ctx->serial);
+				char *activation_token = NULL;
+				if (ctx != NULL) {
+					activation_token = create_xdg_activation_token(ctx->surface,
+						ctx->seat, ctx->serial);
+				}
 				notify_action_invoked(action, activation_token);
 				free(activation_token);
 				break;
 			}
 		}
 		close_notification(notif, MAKO_NOTIFICATION_CLOSE_DISMISSED);
 		break;
+	case MAKO_BINDING_EXEC:
+		assert(binding->command != NULL);
+		pid_t pid = fork();
+		if (pid < 0) {
+			perror("fork failed");
+			break;
+		} else if (pid == 0) {
+			// Double-fork to avoid SIGCHLD issues
+			pid = fork();
+			if (pid < 0) {
+				perror("fork failed");
+				_exit(1);
+			} else if (pid == 0) {
+				char id_str[32];
+				snprintf(id_str, sizeof(id_str), "%" PRIu32, notif->id);
+				setenv("id", id_str, 1);
```

Comment:
I don't feel too strongly either way, but other tools I use that have similar functionality usually namespace the environment variables (like QUTE_WHATEVER in qutebrowser). Would MAKO_ID or something be better just to make it clear where it's coming from since we can't avoid it leaking into child processes?

Reply:
Right. I've been wondering about this above: https://github.com/emersion/mako/pull/366#issuecomment-874190215

I've updated the PR to not use env variables, so it's not leaked to child processes anymore. The variable is now purely local to the shell script. Is that enough to remove your concern about namespacing?

Reply:
Yep, I like this better.